### PR TITLE
referral tracking stats

### PR DESCRIPTION
**Pull request by megaherz**

### CODE REVIEW

File: internal/storage/postgres/postgres.go [hunk 1]
```diff
@@ -129,7 +129,16 @@ func (p pg) GetReferralTrackingByReceiver(ctx context.Context, receiver string)
 	return &r, nil
 }
 
-func (p pg) MarkReferralTrackingInstalled(ctx context.Context, receiver string) error {
+func (p pg) GetReferralTrackingStats(ctx context.Context, sender string) ([]*storage.ReferralTrackingStats, error) {
+	var stats []*storage.ReferralTrackingStats
+	err := sqlx.SelectContext(ctx, p.db, &stats,
+		`SELECT * FROM referral_tracking_sender_stats($1, NULL)	
+	           UNION ALL
+	           SELECT * FROM referral_tracking_sender_stats($1, '30 days'::INTERVAL)`, sender)
```
Comment (mikluke) [inline internal/storage/postgres/postgres.go:137]:
```suggestion
	err := sqlx.SelectContext(ctx, p.db, &stats, `
		SELECT * FROM referral_tracking_sender_stats($1, NULL)	
		UNION ALL
		SELECT * FROM referral_tracking_sender_stats($1, '30 days'::INTERVAL)
	`, sender)
```

File: internal/storage/postgres/postgres.go [hunk 2]
```diff
@@ -141,6 +150,21 @@ func (p pg) MarkReferralTrackingInstalled(ctx context.Context, receiver string)
...
 
+func (p pg) TransitionReferralTrackingToConfirmed(ctx context.Context, receiver string,
+	senderReward, receiverReward int) error {
+	_, err := p.db.ExecContext(ctx,
+		`UPDATE referral_tracking
+               SET status = 'confirmed',
+                   sender_reward = $2,
+                   receiver_reward = $3,
+                   confirmed_at = CURRENT_TIMESTAMP
+                WHERE receiver = $1`, receiver, senderReward, receiverReward)
```
Comment (mikluke) [inline internal/storage/postgres/postgres.go:161]:
```suggestion
	_, err := p.db.ExecContext(ctx, `
		UPDATE referral_tracking
		SET status = 'confirmed',
			sender_reward = $2,
			receiver_reward = $3,
			confirmed_at = CURRENT_TIMESTAMP
                WHERE receiver = $1
	`, receiver, senderReward, receiverReward)
```

File: scripts/migrations/postgres/20211109185301_referral.up.sql [hunk 3]
```diff
@@ -86,4 +86,36 @@ CREATE TABLE referral_tracking (
...
+                    WHERE sender = addr AND installed_at IS NOT NULL AND
+                          CASE WHEN since IS NULL THEN TRUE ELSE registered_at > NOW() - since END), 0)::INT AS installed,
+               COALESCE(
+                   (SELECT COUNT (*) FROM referral_tracking
+                    WHERE sender = addr AND confirmed_at IS NOT NULL AND
+                          CASE WHEN since IS NULL THEN TRUE ELSE registered_at > NOW() - since END), 0)::INT AS confirmed,
+               COALESCE(
+                   (SELECT SUM (COALESCE(sender_reward, 0)) FROM referral_tracking
+                   WHERE sender = addr AND
+                         CASE WHEN since IS NULL THEN TRUE ELSE registered_at > NOW() - since END), 0)::INT AS reward;
```
Comment (mikluke) [inline scripts/migrations/postgres/20211109185301_referral.up.sql:117]: It would be great to format the code a bit

Comment (mikluke) [reply scripts/migrations/postgres/20211109185301_referral.up.sql:117]: also what do you think about to add `f_` prefix to the function's name?

**Review by mikluke — APPROVED** [review]
Just some cosmetic stuff

File: internal/storage/postgres/postgres.go (hunk 2 above)
Comment (megaherz) [reply internal/storage/postgres/postgres.go]: formatted

File: scripts/migrations/postgres/20211109185301_referral.up.sql (hunk 3 above)
Comment (megaherz) [reply scripts/migrations/postgres/20211109185301_referral.up.sql]: formatted.
Regarding `_f` I don't think we need it. There are no places with context when you are not able to identify where it is a func name or not.

**Pull request merged**